### ISSUE TITLE
PiecesOS using too much RAM/spike in CPU

### ISSUE THREAD

**Issue by rosie-at-pieces** [pieces-app/support#150]
Several users have called in recently about Pieces using too much RAM or causing a spike in CPU up to 100%. We are trying to track down this issue, and would appreciate any feedback about your experience with this if you are experiencing it. 

We have a hunch it has something to do with internet connections, but aren't fully sure yet. 

If you're experiencing this and wouldn't mind hopping on a call with us to show us, we would really appreciate it to help us troubleshoot! You can schedule a call [here](https://calendar.google.com/calendar/u/0/appointments/schedules/AcZssZ2iaRWpd1kA2k_xg2KrVi_jiTHo0POiMZSI9WAYsxs5jJ9sU2Rd2zKitLuSFd7l3FhatGzy5Vgt
). 

A short-term solution is to restart PiecesOS whenever this occurs.

**Comment by mark-at-pieces** [comment]
note: when running a CPU || GPU  LLM (w/in the Pieces Copilot) this is to be expected.

however, for anyother case this is 100% no expected behavior and I believe that we have a fix for this but will need to track this over a bit of time to see if we have any sort of regression w/ regard to CPU and memory

https://github.com/pieces-app/os_server/issues/676

**Comment by fahad-programmer** [comment]
Well in my case pieces os is taking so much ram and when i click on optimize memory usage it cuts down 1gb of ram usage and then immediately start using more ram then ever before

**Comment by mark-at-pieces** [comment]
> Well in my case pieces os is taking so much ram and when i click on optimize memory usage it cuts down 1gb of ram usage and then immediately start using more ram then ever before

Hey @fahad-programmer sorry your getting some issues w/ regard to your ram:

Q: what is the current memory usage that your PiecesOS instance is at right now?

Also is your memory continue to scale or is it stable? at roughly 1.6gbs?

**Comment by fahad-programmer** [comment]
Its actually talking close to 5gb or ram and i am not even using it. its running in background looks like they are mining using this software something.

**Comment by mark-at-pieces** [comment]
@fahad-programmer sorry for the delayed response,

want to ensure we solve what every issue your getting w/ memory

here is a link to my calendar and hopefully I can solve whatever might be going wrong with your POS instance.

[link](https://calendar.app.google/jwHCZ2WdnY89BwM98)

TLDR:
POS is a centralized server that runs locally on your machine and connects all the Pieces product suite.

so basically w/o POS you cannot do anything.

POS memory sources:
- ML models + local processing
- Streaming information to products from the other products
- r/w/u to the db

Note: if you are running at 5GB there 100% is a memory leak that I would love to help diagnose and resolve.
Note: we are 100% not mining

**Comment by ayush-rudani** [comment]
I am facing a similar issue on my Linux machine (Ubuntu 24.04), where the memory usage of the app is very high, even under normal conditions (around 2+ GB) and sometimes exceeding 3 GB. (No models are installed or running).

**Comment by qomosoloto** [comment]
Not too much RAM used but with higher CPU in windows.

**Comment by CodeConnoisseur74** [comment]
I am running Ubuntu 22.04 and experiencing the same issues as @ayush-rudani . @mark-at-pieces @rosie-at-pieces

**Comment by shldk** [comment]
Tried this long time ago, ditched due to it using almost all of my RAM, uninstalled. Now, I think a year passed, I decided to give it another go hoping things are normal now, but nah, just installed it, did nothing at all, it's just sitting in background using 1.8GB of RAM. What's strange to me is everyone suggests it and somehow doesn't notice or?.. Sorry, unfortunately have to uninstall again :(

**Comment by verdant-spark** [comment]
Chuck my tuppenceworth; I really like Pieces, and it's changed the way that I work, but I now need to start it up, use it, and then shut it down again as it consumes 50% of my CPU all the time and slows _everything_ down.  Happy to jump on a call to troubleshoot if that's useful.

**Comment by fahad-programmer** [comment]
Guys i have figured out why this happens as u use live context the pieces tries to keep track of everything u do while developing so it knows many things beforehand that is why it uses much ram so if u want to consume less ram try turing off the live context.

**Comment by apetta** [comment]
> Hi everyone! Thanks for all the feedback; this is an older ticket but it's still relevant especially as we've added a lot of features since April, particularly the ability to use [Live Context](https://docs.pieces.app/product-highlights-and-benefits/live-context), which is powered by the [Workstream Pattern Engine](https://docs.pieces.app/product-highlights-and-benefits/live-context#the-workstream-pattern-engine).
> 
> At this moment, it is fairly standard for PiecesOS to use ~2GB of resources, although we are working constantly to make this more performant.
> 
> For context (you all know we love context here at Pieces), building **on-device** ML is fundamentally difficult. We're one of only a handful of companies focused on truly secure, private, and on-device ML/AI capabilities and in the early days it has been at the cost of memory consumption. That said, we have lots of improvements on this around the corner if you're willing to hang tight!
> 
> In the meantime, you can click on the Pieces "P" icon in the global macOS toolbar/Windows taskbar to get a popover menu that has an option to "Optimize Memory." This will unload some of the current in-use ML models and hopefully free up some RAM.
> 
> Thanks to EVERYONE for being early adopters and for your patience as we try and solve some of these on-device ML constraints. If you are experiencing more than ~2GB of usage, let's hop on a call so we check it out - use [this link](https://calendar.app.google/uEWkjrpgvKxqsSpm8)!

The heavy RAM usage has meant that I've had to stop using Pieces for now. Hopefully it has a lighter footprint in the near future!

**Comment by tiptronic** [comment]
Just as a heads up: I'm still deactivating Pieces OS on a daily basis, because of CPU spikes of 120% or more on my:

Apple M1 Max, 32GB RAM, MacOS Sequoia 15.2 (24C98)

**Comment by curtis2point1** [comment]
Similar issues on my end.  Just running in the background the Pieces OS is causing spikes up to 50% of CPU and slowing down my whole system.  I haven't been able to identify any triggers or clear correlations.

Can't wait for this to get fixed as I absolutely love the tool and the interface.  It's going to be a game changer once the footprint is reasonable.

![Image](https://github.com/user-attachments/assets/e4345fdc-8b0f-4861-b36d-f22db79aa5d1)

**Comment by jgeofil** [comment]
I've been having the same issue, and I can't identify what triggers it. Disabling long-term memory makes no difference. It gets to the point my whole system is practically frozen and I have to shut it down.[]
![](https://imagizer.imageshack.com/a/img923/4927/FWOQZM.png)

**Comment by mark-at-pieces** [comment]
Hey Everyone, 

We have a release going out tomorrow or Thursday for PiecesOS 11.1.0, this release is replacing 2 heavy database technologies that we believe were causing memory and cpu issues.

With 11.1.0, this should help the cpu usage, however, we are actively working on 2 additional fixes to improve this in the near future.

Please let us know if this 11.1.0 release reduces cpu + memory, we appreciate all the help and feedback from the community you all are awesome!!!!

**Comment by Corbanw** [comment]
I noticed my computer was slowing down a lot, so i opened task manager to check on things and my cpu usage was at a steady 100%. I scrolled down in the programs to find that pieces os was using most of it. After i launched the app, it ran a few updates then everything went back to normal. Pieces os was using above 85 percent of the cpu before it updated.

**Comment by curtis2point1** [comment]
The update seems to have made a dramatic improvement on my system so far.  CPU usage dropped from 30-50% down to 3-5%.

It's still one of the more resource intensive apps that I have running in the background; however, I expect that given all the functionality ...and, most importantly, it's not impacting system usability now.

<img width="398" alt="Image" src="https://github.com/user-attachments/assets/c51923e0-c10c-4b82-9fad-baa778ed05eb" />

**Comment by apetta** [comment]
I do wish you guys had a 'lite' mode where all these extra features are turned off. I would just like somewhere to store snippets, with some light AI-powered context enhancement - without taking over my machine.

I am just using Apple notes at this point...

**Comment by ryanm-ill-er** [comment]
I set pieces to only use my E-cores, my temperatures were getting too high. Do that in task manager/system explorer's process affinity.

**Comment by chrstphr84** [comment]
I see this issue is still open, but is it still being investigated after the release of 11.1.0 a couple weeks ago and LTM-2 this week?  I am hoping to keep attention on the issue, as I am consistently seeing 100% usage CPU from Pieces OS and extremely high virtual memory usage after the LTM-2 update.  

M3 Pro MacBook for context.  
<img width="671" alt="Image" src="https://github.com/user-attachments/assets/ce32e5a5-9d12-4309-b232-45c3dc5555f0" />

**Comment by tiptronic** [comment]
Jep - me too:

<img width="435" alt="Image" src="https://github.com/user-attachments/assets/d8db23f8-7aa3-411f-a86d-7c7aa6fc40b3" />

**Comment by Kepners** [comment]
I am getting a huge load on my system. I have two instances of PIeces installed. Nothing i do changes this loading. Eventually, everything crashed, and i have Pieces only watching VSCODE... But its still unable to connect within VSCODE.

<img width="839" height="313" alt="Image" src="https://github.com/user-attachments/assets/e666703d-7b44-4623-ab11-f52f727a1ad4" />

![Image](https://github.com/user-attachments/assets/9b325cbd-0f4b-4d95-8fa4-8256c9f0eac9)

<img width="437" height="641" alt="Image" src="https://github.com/user-attachments/assets/cee459ea-6187-4d8d-8fd6-43a5f34d1b3a" />

**Comment by bugparty** [comment]
> [@Kepners](https://github.com/Kepners)
> 
> This is not good!
> 
> How did you install Pieces OS? Seems to be multiple things going on here that I don't quite understand.
> 
> The latest versions should be:
> 
> Pieces OS: 12.2.1 Pieces Desktop App: 4.3.4
> 
> I would go ahead and uninstall Pieces OS as well as the "Unknown Origin" application, then reinstall Pieces OS through this link:
> 
> https://builds.staging.pieces.app/stages/production/appinstaller/os_server.appinstaller

I encountered a memory leak issue that piecesOS will use up my 128gb ram, I am not sure if it is because I am using intel core ultra2 cpu and windows11, I remember weeks' ago it doesn't have memory leak issue.

**Comment by Kepners** [comment]
I had to uninstall Pieces completely. The reinstalled it to get the ram to resolve itself. I have found further issues with piceces - VScode not integrating with Pieces and the MCP is so hit and miss, when it work there are no summary titles added. Its getting frustrating using this.

**Comment by thezaidsheikh** [comment]
In my case also it is the highest memory and cpu usage service running
